### PR TITLE
Add option historyApiFallback

### PR DESCRIPTION
**Pull request by drudrum**
This PR contains a:

- [ ] **bugfix**
- [ x] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

Adds support of history api (Example: React Router5)
Like this https://webpack.js.org/configuration/dev-server/#devserverhistoryapifallback

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Why?

**Comment by drudrum** [comment]
I want use react router5 in my app. 
This is not breaking change.  

i want add option like as
https://webpack.js.org/configuration/dev-server/#devserverhistoryapifallback

**Comment by alexander-akait** [comment]
You can do it on own side, why we need add it here?

**Comment by drudrum** [comment]
Many people want to see compatibility with webpack-dev-server options (i think). 
I have not direct and clear access to memfs from my app. 
3 rows in library VS 50 rows in each react app using router5

**Comment by alexander-akait** [comment]
Please provide example of the problem, because you can use `connect-history-api-fallback` on own application side, I don't see any problems in our bug trackers from other developers

**Comment by alexander-akait** [comment]
Can you create simple git repo? I think I understand the problem, but I would like to check this before merge? Thanks

**Comment by drudrum** [comment]
https://github.com/drudrum/exampleRouter5

**Review by alexander-akait — COMMENTED** [review]
ah, you want to read `index.html` from memory fs

We should add docs and tests on this option, please fix it and I will merge

**Comment by drudrum** [comment]
Readme at https://github.com/webpack/webpack-dev-middleware/pull/922/commits/59382db08e47b9961153930d24b6b38f9765f60f

**Comment by drudrum** [comment]
How i can fix it?

**Comment by alexander-akait** [comment]
Please rebase, should work

**Comment by drudrum** [comment]
Workflow dislike me ))
How i can fix this problem?

File: types/utils/setupWriteToDisk.d.ts [hunk 1]
```diff
@@ -11,20 +11,14 @@ export = setupWriteToDisk;
...
-type Compiler = import("webpack").Compiler;
-type MultiCompiler = import("webpack").MultiCompiler;
-type Compilation = import("webpack").Compilation;
-type IncomingMessage = import("../index.js").IncomingMessage;
-type ServerResponse = import("../index.js").ServerResponse;
+type Compiler = import('webpack').Compiler;
+type MultiCompiler = import('webpack').MultiCompiler;
+type Compilation = import('webpack').Compilation;
+type IncomingMessage = import('../index.js').IncomingMessage;
+type ServerResponse = import('../index.js').ServerResponse;
```
Comment (alexander-akait) [inline types/utils/setupWriteToDisk.d.ts:24]: Looks like we need to run prettier on all types files, can you do it?

Comment (drudrum) [reply types/utils/setupWriteToDisk.d.ts:24]: Yes of course

**Comment by alexander-akait** [comment]
@drudrum Friendly ping

**Comment by drudrum** [comment]
@alexander-akait Sorry. We should finish this old PR :-)

**Comment by drudrum** [comment]
@alexander-akait Can you find approvers for this PR?

File: src/utils/getFilenameFromUrl.js [hunk 2]
```diff
@@ -95,8 +95,15 @@ function getFilenameFromUrl(context, url) {
         filename = path.join(outputPath, querystring.unescape(pathname));
       }
 
-      let fsStats;
+      if (
+        context.outputFileSystem.existsSync &&
+        !context.outputFileSystem.existsSync(filename) &&
+        options.historyApiFallback
+      ) {
+        filename = path.join(outputPath);
```
Comment (alexander-akait) [inline src/utils/getFilenameFromUrl.js:103]: Why do we need `path.join` here, it doesn't have sense

Comment (drudrum) [reply src/utils/getFilenameFromUrl.js:103]: I will check. Look like mistake. May be we need change it to path.resolve

Comment (drudrum) [reply src/utils/getFilenameFromUrl.js:103]: I think it is required.
image

May be i can remove condition?

Comment (drudrum) [reply src/utils/getFilenameFromUrl.js:103]: Now this code looks like this option is not necessarily at all.

Comment (drudrum) [reply src/utils/getFilenameFromUrl.js:103]: Everything should work without this option.
I am confused. 
I feel that i need cancel this PR.

**Comment by drudrum** [comment]
This is duplicated PR. We have the same #930

**Pull request closed without merging**